### ISSUE TITLE
Label problem

### ISSUE THREAD

**Issue by jasonyeh-dev** [arturjordao/WearableSensorData#5]
@arturjordao Hello!
I am training the model using your preprocessed dataset.(WISDM.npz, mHealth.npz...)
I was confused when making the confusion matrix.
Because I don't know the labels belong to which activities.
Could you give me a hint!
Thanks in advance!

**Comment by arturjordao** [comment]
Dear @108618026,

I hope the attached file helps you. In this file, you send a dataset name as input and receive a dictionary. Each key in the dictionary indicates one label, e.g., 10: 'Running' means that the y=10 is the activity running.
[activity_name.zip](https://github.com/arturjordao/WearableSensorData/files/8530853/activity_name.zip)

Best regards,
Artur Jordão.

**Comment by jasonyeh-dev** [comment]
Dear @arturjordao,

Thanks for your reply.
From the file that you attached. The relationship of  labels and activities in WISDM is look like the following.

actNameWISDM = {
			0: 'Jogging',
			1: 'Walking',
			2: 'Upstairs',
			3: 'Downstairs',
			4: 'Sitting',
			5: 'Standing'
		}
And I use the following codes to see how many counts in the activities, It seems like the label number 5 has the most counts. Its not reasonable because from the author of WISDM dataset(https://www.cis.fordham.edu/wisdm/dataset.php), the activity of **walking** should be the most counts(walking>jogging>upstairs>downstairs>sitting>standing), but from your attached file, label number 5 is standing not walking.
Please confirm that for me!
best regard!
Jason
  
tmp = np.load("WISDM.npz")
test = np.argmax(tmp["y"],axis = 1)
vals,counts = np.unique(test, return_counts=True)
print(counts) #[1824 6522 1161  933 2265 8141]
print(vals) #[0 1 2 3 4 5]

**Comment by arturjordao** [comment]
Dear @108618026 ,

The label number can differ from the original dataset due to implementation details. Regarding the number of samples (i.e., the counts), the sliding window process probably removed some samples that do not fit in the window size (details in the paper). Therefore, the number of samples for each category can also differ from the original dataset. Finally, you can try to rebuild the dataset, for this purpose, please look at the closed issue #2.

I hope this helps you.

Best regards, 
Artur Jordão.

**Comment by jasonyeh-dev** [comment]
@arturjordao 

I appreciate your reply. I will try to rebuild the dataset.

Thank you!

Best regards,
Jason

**Comment by ZQhaha** [comment]
Dear @108618026 ,
I have modified the original data spliting code for generating the label. Maybe it can do some help.
```python
import numpy as np
import glob
import os

def save_file(act, subject, trial_id, trial, output_dir):
    output_file_name = '{}_s{}_t{}.txt'.format(act.lower(), subject, trial_id)
    output_file_name = os.path.join(output_dir,output_file_name)
    np.savetxt(X=trial, fmt='%s', fname=output_file_name, delimiter=' ', )

if __name__ == '__main__':
    file_name = 'WISDM_ar_v1.1/WISDM_ar_v1.1_raw.txt'
    output_dir = 'wisdm_raw'
    if not os.path.exists(output_dir):
        os.mkdir(output_dir)
    separator = '_'
    idx_label = 1
    f = open(file_name)
    String = f.read()
    String = String.replace(';','\n')
    lines = [s for s in String.splitlines() if s]
    iterator = 0
    trial = []
    trial_id = 0
    count = 0
    for line in lines:
        count += 1
        split = line.strip().split(',')
        try:
            subject, act, time_stamp, x, y, z = line.rstrip(',').split(',')
        except:
            subject, act, time_stamp, x, y = line.rstrip(',').split(',')
            z = 0
            print(line)

        #It is the same trial
        if iterator < len(lines)-1 and lines[iterator+1].split(',')[1] == act and lines[iterator+1].split(',')[0] == subject:
                if time_stamp != '0': #timestamp equal to zero is a bug of the trial
                    trial.append((x, y, z))

        #The next line will be a novel trial
        else:
            print('{} {} {}'.format(act, subject, trial_id))
            save_file(act, subject, trial_id, trial, output_dir)
            trial_id = trial_id + 1
            trial = []
        #print(subject, act, x, y, z)
        iterator = iterator + 1
    f.close()

```
```python
import numpy as np
import random
import os
import shutil
from sklearn.model_selection import LeaveOneGroupOut
from scipy.interpolate import interp1d
import shutil
import moxing as mox
import pandas as pd

def sw(trial=None, size=None, overlapping=None):
    r = 0
    delta = size
    output = []

    sample = trial

    while r+delta < len(sample):
        block = sample[r:r+delta]
        x=np.arange(0,200,2)
        Interp = interp1d(x, block, axis=0,kind='linear',fill_value="extrapolate")
        x_new=np.arange(200)
        y_new=Interp(x_new)
        output.append(y_new)
        r = r+delta
        r = r- (int(delta*overlapping))
    return output

def cv_generator(y, n_folds=10):
    from sklearn.cross_validation import StratifiedKFold
    skf = StratifiedKFold(y, n_folds, shuffle=True)
    folds = []
    for i, (train_idx, test_idx) in enumerate(skf):
        folds.append((train_idx, test_idx))
    return folds

def label_generator(files=None):
    activity = {}
    label_idx = -1
    for file in files:
        label = file.split(separator)[idx_label]
        if label not in activity.keys():
            label_idx = label_idx + 1
            activity[label] = label_idx
    return activity

def subject_trials(files=None):
    subject = {}
    subject_idx = -1
    for file in files:
        idx = file.split(separator)[idx_subject]
        if idx not in subject.keys():
            subject_idx = subject_idx + 1
            subject[idx] = subject_idx
    return subject

def data_generator(files=None, activity=None, individuals_id=None):
    X = []
    y = []
    fundamental_matrix = np.zeros((len(activity),len(individuals_id)))#Matrix Activity (row) by Subject (col)
    groups = []

    for file in files:
        if file.startswith("."):
            continue
        label = file.split(separator)[idx_label]
        subject = file.split(separator)[idx_subject]
        label_idx = activity[label]
        subject_idx = individuals_id[subject]

        file_name = '{}/{}'.format(input_dir, file)
        trial = np.genfromtxt(file_name, delimiter=' ', filling_values=np.nan,
                              case_sensitive=True, deletechars='',replace_space=' ')

        samples = sw(trial=trial, size=sw_size, overlapping=0.5)
        for i in range(0, len(samples)):
            X.append([samples[i]])
            y.append(label_idx)
            groups.append(subject_idx)
            fundamental_matrix[label_idx][subject_idx] += 1

    groups = np.array(groups)
    return X, y, groups, fundamental_matrix

def _to_categorical(y, nb_classes=None):
    '''Convert class vector (integers from 0 to nb_classes)
    to binary class matrix, for use with categorical models
    '''
    if not nb_classes:
        if 0 in y:
            nb_classes = np.max(y) + 1
        else:
            nb_classes = np.max(y)
    Y = np.zeros((len(y), nb_classes))
    for i in range(len(y)):
        Y[i, y[i]] = 1.
    return Y


input_dir = 'wisdm_raw'
file_name = 'wisdm'
separator = '_'
idx_label = 0
idx_subject = 1
sw_size = 100
if os.path.exists('wisdm_raw/.ipynb_checkpoints'):
    shutil.rmtree('wisdm_raw/.ipynb_checkpoints')
files = os.listdir(input_dir)
activity = label_generator(files=files)
individuals_id = subject_trials(files=files)

X, y, groups, fundamental_matrix = data_generator(files=files, activity=activity, individuals_id=individuals_id)

invalid_rows = []
for row in fundamental_matrix:
    check_zeros = np.where(row != 0.)
    if check_zeros[0].shape[0] < 2: #An activity is performed just by one subject
        invalid_rows.append(row)
subject_dict={value:key for key,value in individuals_id.items()}
activity_dict={value:key for key,value in activity.items()}
df=pd.DataFrame(fundamental_matrix,columns=[subject_dict[i] for i in range(len(individuals_id))],
               index=[activity_dict[i] for i in range(len(activity_dict))])
df.to_csv('wisdm_description.csv')
if(len(invalid_rows) == 0):
    loso = LeaveOneGroupOut()
    tmp = loso.split(X=X, y=y, groups=groups)
    folds =[]
    for train_index, test_index in loso.split(X, y, groups):
        folds.append((train_index, test_index))

    X = np.array(X)
    y = np.array(y)
    y = _to_categorical(y, len(activity))
    np.savez_compressed(file_name, X=X, y=y, folds=folds)
else:
    print('Problem at lines')
    for row in invalid_rows:
        print(row)
```

**Comment by jasonyeh-dev** [comment]
Dear @ZQhaha, 

Thank you very much. I will try your python codes as soon as possible.
